### ISSUE TITLE
Deviation from reghdfe std errors with multiway & interacted clustering

### ISSUE THREAD

**Issue by maxnorton** [FixedEffects/FixedEffectModels.jl#49]
@yywingliang and I identified the cause of the divergence between the standard errors generated for [these multiway clustered se test models](https://github.com/matthieugomez/FixedEffectModels.jl/blob/7eb799769531d60f7b1da882f0fbc2ace2c52175/test/reg.jl#L256-L268) and the results given for the same models in Stata using reghdfe. Thought I would detail it here because the current algorithm seems deliberately chosen, so I wanted to check whether it would be most helpful for us to document but leave code as is, submit a PR overwriting the current algorithm with the reghdfe approach, or somehow add an optional argument switching between the two.

The current algorithm in src/vcov/vcovcluster.jl estimates the variance-covariance matrix as the sum of several component variance estimates which are individually [weighted by C / C - 1](https://github.com/matthieugomez/FixedEffectModels.jl/blob/7eb799769531d60f7b1da882f0fbc2ace2c52175/src/vcov/vcovcluster.jl#L78-L82), where C is the number of clusters in the dimension of the component, only if the number of clusters in that dimension is less than the number of observations.

In contrast, reghdfe weights the sum of the component variances by C' / C' - 1, where C' is the number of clusters in the smallest cluster dimension.

Would it be preferable to document the difference, replace with the reghdfe algorithm, or implement the reghdfe algorithm as an alternative option (and if the last, is there an approach you'd like to see us follow)?

Thanks!

**Comment by matthieugomez** [comment]
Great sleuthing is work! That’s very nice.
I am not sure which one is preferable though.
Could you find justifications for either / both in the litterature? Could
you check what ivreg2 does in the case without fixed effects?

**Comment by matthieugomez** [comment]
You may want to check the discrepancy with clustered standard error and
ivreg2 too
https://github.com/matthieugomez/FixedEffectModels.jl/blob/7eb799769531d60f7b1da882f0fbc2ace2c52175/test/reg.jl#L392.
in case the discrepancy is related.

**Comment by maxnorton** [comment]
Both bias corrections are described as typical in Cameron Gelbach & Miller (2011; section 2.3), though they use the component-by-component correction in their applied exercises in that paper and in their cgmreg stata command. I haven't encountered an explicit theoretical argument for one over the other anywhere. I guess worth considering that using the minimum cluster dimension approach is more conservative than the cluster-by-cluster approach, giving a larger std error. 

According to ivreg2 documentation that package uses the minimum approach, like reghdfe. But in practice, the standard errors in ivreg2 are significantly smaller, and I'm not really sure why. @yywingliang and I noticed that the standard errors for the models in test/reg.jl lines [212](https://github.com/matthieugomez/FixedEffectModels.jl/blob/7eb799769531d60f7b1da882f0fbc2ace2c52175/test/reg.jl#L212) and [226](https://github.com/matthieugomez/FixedEffectModels.jl/blob/7eb799769531d60f7b1da882f0fbc2ace2c52175/test/reg.jl#L226) require a -2 df adjustment to match the ivreg2 standard errors. But even the combination of the -2 df adjustment and the minimum cluster adjustment doesn't bring the standard error in the model you mentioned (line 392) into line, so there's something else still unidentified going on.

**Comment by matthieugomez** [comment]
Thanks for the detailed answer. Let us do like reghdfe then. Can you add a comment in the code documenting the change?

**Comment by matthieugomez** [comment]
Also, for your pull request, could you update the tests with comments indicating the commands that now match Stata and the ones that still do not? (including the ones corrected by your previous pull request)

**Comment by maxnorton** [comment]
[Pull request sent!](https://github.com/matthieugomez/FixedEffectModels.jl/pull/50) Re your requests: the removal of the old error adjustment and the implementation of the new one are commented in src/vcov/vcovcluster.jl. For each of the tests in the standard errors section of test/reg.jl (including 2 new tests), I indicated whether or not they match Stata (most do, except a couple of IV tests). There appears to be a continuing issue with IVs relative to Stata that is not fully fixed here (though some cases of ivreghdfe match exactly). I had to make some changes to the test values of 1st stage F stats (approx line 403 in test/reg.jl) that were the product of the change to multiway clustering. Again with that section, commented which do and do not completely match Stata. It looks like there may still be multiple interacting issues contributing to the divergence of that statistic.
